CrossSectionAnalysis/cpp/CrossSectionCompute.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * This file was created by SET [Surgeon] [Hobbyist developer].
+ */
+
+#ifndef __CrossSectionCompute_h
+#define __CrossSectionCompute_h
+
+/*
+ * Notes for me :
+ * Only class names must be 'vtk' prefixed.
+ */
+
+// Created by cmake
+#include "vtkSlicerCrossSectionAnalysisModuleLogicExport.h"
+#include <thread>
+
+#include <vtkDoubleArray.h>
+#include <vtkMRMLNode.h>
+#include <vtkPolyData.h>
+#include <vtkSmartPointer.h>
+#include <vtkObjectFactory.h>
+#include <sys/time.h>
+
+#define DEVTIME 1
+
+/**
+ * This class computes cross-section areas along a centerline
+ * in joinable threads running in parallel.
+*/
+class VTK_SLICER_CROSSSECTION_COMPUTE_EXPORT vtkCrossSectionCompute
+: public vtkObject
+{
+public:
+
+  static vtkCrossSectionCompute *New();
+  vtkTypeMacro(vtkCrossSectionCompute, vtkObject);
+  void PrintSelf(ostream& os, vtkIndent indent) override;
+
+  void SetNumberOfThreads(unsigned int number)
+  {
+    numberOfThreads = number;
+  }
+  void SetInputCenterlineNode(vtkMRMLNode * inputCenterline)
+  {
+    inputCenterlineNode = inputCenterline;
+  }
+  /**
+   * inputSurface may be a segmentation or a model.
+   * inputSegmentId is ignored if inputSurface is a model.
+   * A member closed surface polydata is derived from the input.
+  */
+  bool SetInputSurfaceNode(vtkMRMLNode * inputSurface, const std::string& inputSegmentId);
+  
+  /**
+   * This is the main purpose of this class.
+   * The cross-section area and circular equivalent diameter
+   * columns of the output table are updated in parallel.
+   */
+  void UpdateTable(vtkDoubleArray * crossSectionAreaArray, vtkDoubleArray * ceDiameterArray);
+  
+private:
+  unsigned int numberOfThreads = 1;
+  vtkMRMLNode * inputCenterlineNode = NULL;
```

Comment:
This is unsafe. The inputCenterlineNode will be deleted as soon as the caller does not need the object anymore. Instead, you can use a vtkSmartPointer<vtkMRMLNode>, which increments the reference counter (preventing deletion) until this vtkCrossSectionCompute class needs the node.